ledger/participant-integration-api/src/main/scala/platform/index/IndexServiceImpl.scala

```diff
@@ -90,29 +93,42 @@ private[index] class IndexServiceImpl(
       filter: domain.TransactionFilter,
       verbose: Boolean,
   )(implicit loggingContext: LoggingContext): Source[GetTransactionsResponse, NotUsed] =
-    between(startExclusive, endInclusive) { (from, to) =>
-      from.foreach(offset =>
-        Spans.setCurrentSpanAttribute(SpanAttribute.OffsetFrom, offset.toHexString)
-      )
-      to.foreach(offset =>
-        Spans.setCurrentSpanAttribute(SpanAttribute.OffsetTo, offset.toHexString)
-      )
-      dispatcher()
-        .startingAt(
-          from.getOrElse(Offset.beforeBegin),
-          RangeSource(transactionsReader.getFlatTransactions(_, _, convertFilter(filter), verbose)),
-          to,
+    withValidatedFilter(filter)(
+      between(startExclusive, endInclusive) { (from, to) =>
+        from.foreach(offset =>
+          Spans.setCurrentSpanAttribute(SpanAttribute.OffsetFrom, offset.toHexString)
         )
-        .map(_._2)
-        .buffered(metrics.daml.index.flatTransactionsBufferSize, LedgerApiStreamsBufferSize)
-    }.wireTap(
-      _.transactions.view
-        .map(transaction =>
-          Event(transaction.commandId, TraceIdentifiers.fromTransaction(transaction))
+        to.foreach(offset =>
+          Spans.setCurrentSpanAttribute(SpanAttribute.OffsetTo, offset.toHexString)
         )
-        .foreach(Spans.addEventToCurrentSpan)
+        val memoizedConvertedFilter = memoizedFilterRelationAndInterfaceProjection(filter)
```

Comment:
By reusing the same filter transformation for each `(startExclusive, endInclusive)` chunk, you are missing out on a potential optimization: the set of templates that implement a given interface depends on the offset. If you know that a template was introduced after `endInclusive` (or deleted before `startInclusive` if we ever start deleting packages), you don't need to include it in the filter for that offset range.

I assume this is less important than reducing the cost of doing the filter transformation?

Reply:
In the current stage we do not have deletions, and therefore:
* in case of tailing streams we are anyway will have all the possible templates. pls note this is executed for each IMFO buffer update (very frequent)
* in persistente case for old start and old end dates you are right, but those are historical queries, so this map building will be executed once: not so important to optimize anyway